File: helpers/categoryMatcher.js
function findSimilarTransactionCategory(description, existingTransactions, maxResults = 100) {
  if (!description || !existingTransactions || existingTransactions.length === 0) {
    return null;
  }

  const descLower = description.toLowerCase().trim();

  const recentTransactions = existingTransactions
    .sort((a, b) => new Date(b.date) - new Date(a.date))
    .slice(0, maxResults)
    .filter(t => t.category && t.category !== 'Uncategorized');

  if (recentTransactions.length === 0) {
    return null;
  }

  const similarities = recentTransactions.map(transaction => {
    const transactionDesc = transaction.description.toLowerCase().trim();
    const similarity = calculateDescriptionSimilarity(descLower, transactionDesc);
    return {
      transaction,
      similarity,
      category: transaction.category
    };
  });

  const goodMatches = similarities
    .filter(match => match.similarity > 0.3)
    .sort((a, b) => b.similarity - a.similarity);

  if (goodMatches.length === 0) {
    return null;
  }

  const categoryScores = new Map();
  goodMatches.forEach(match => {
    const category = match.category;
    if (!categoryScores.has(category)) {
      categoryScores.set(category, { total: 0, count: 0, maxSimilarity: 0 });
    }
    const score = categoryScores.get(category);
    score.total += match.similarity;
    score.count += 1;
    score.maxSimilarity = Math.max(score.maxSimilarity, match.similarity);
  });

  let bestCategory = null;
  let bestScore = 0;

  categoryScores.forEach((score, category) => {
    const avgSimilarity = score.total / score.count;
    const weightedScore = (avgSimilarity * 0.7) + (score.maxSimilarity * 0.3);

    if (weightedScore > bestScore) {
      bestScore = weightedScore;
      bestCategory = category;
    }
  });

  return bestScore > 0.4 ? bestCategory : null;
}

function calculateDescriptionSimilarity(desc1, desc2) {
  if (desc1 === desc2) return 1.0;

  const words1 = new Set(desc1.split(/\s+/).filter(w => w.length > 2));
  const words2 = new Set(desc2.split(/\s+/).filter(w => w.length > 2));

  if (words1.size === 0 || words2.size === 0) return 0;

  const intersection = new Set([...words1].filter(x => words2.has(x)));
  const union = new Set([...words1, ...words2]);
  const jaccardSimilarity = intersection.size / union.size;

  let substringScore = 0;
  const words1Array = Array.from(words1);
  const words2Array = Array.from(words2);

  for (const word1 of words1Array) {
    for (const word2 of words2Array) {
      if (word1.includes(word2) || word2.includes(word1)) {
        substringScore += 0.5;
      }
    }
  }
  const normalizedSubstringScore = Math.min(substringScore / Math.max(words1.size, words2.size), 1);

  let merchantScore = 0;
  const commonMerchants = ['amazon', 'walmart', 'target', 'starbucks', 'mcdonalds', 'uber', 'lyft', 'netflix', 'spotify'];
  const desc1Lower = desc1.toLowerCase();
  const desc2Lower = desc2.toLowerCase();

  for (const merchant of commonMerchants) {
    if (desc1Lower.includes(merchant) && desc2Lower.includes(merchant)) {
      merchantScore += 0.3;
    }
  }

  const finalScore = (jaccardSimilarity * 0.5) + (normalizedSubstringScore * 0.3) + (merchantScore * 0.2);

  return Math.min(finalScore, 1.0);
}

module.exports = { findSimilarTransactionCategory, calculateDescriptionSimilarity };

File: backend/app/utils/category_matcher.py
"""
Category matching via description similarity.

Ports legacy/helpers/categoryMatcher.js — Jaccard similarity + substring
matching + merchant recognition for auto-categorizing imported transactions.
"""

KNOWN_MERCHANTS = [
    "amazon",
    "walmart",
    "target",
    "costco",
    "starbucks",
    "mcdonalds",
    "uber",
    "lyft",
    "netflix",
    "spotify",
    "apple",
    "google",
    "venmo",
    "paypal",
    "chase",
    "wells fargo",
    "bank of america",
]


def calculate_description_similarity(desc1: str, desc2: str) -> float:
    """Calculate similarity between two transaction descriptions.

    Scoring components (weighted):
    - Jaccard word similarity: 40%
    - Substring/prefix matching: 30%
    - Known merchant matching: 30%

    Returns float 0.0–1.0.
    """
    d1 = desc1.strip().lower()
    d2 = desc2.strip().lower()

    if d1 == d2:
        return 1.0

    # Jaccard similarity on words > 2 chars
    words1 = {w for w in d1.split() if len(w) > 2}
    words2 = {w for w in d2.split() if len(w) > 2}

    if not words1 and not words2:
        return 1.0

    if not words1 or not words2:
        jaccard = 0.0
    else:
        intersection = words1 & words2
        union = words1 | words2
        jaccard = len(intersection) / len(union)

    # Substring scoring — word prefix/suffix matches
    substring_score = 0.0
    if words1 and words2:
        matches = 0
        total = len(words1)
        for w1 in words1:
            for w2 in words2:
                if w1.startswith(w2) or w2.startswith(w1):
                    matches += 1
                    break
                if w1.endswith(w2) or w2.endswith(w1):
                    matches += 0.5
                    break
        substring_score = matches / total if total > 0 else 0.0

    # Merchant matching
    merchant_score = 0.0
    for merchant in KNOWN_MERCHANTS:
        if merchant in d1 and merchant in d2:
            merchant_score = 1.0
            break

    final = jaccard * 0.4 + substring_score * 0.3 + merchant_score * 0.3
    return min(final, 1.0)


def find_similar_category(
    description: str,
    existing_transactions: list[dict],
    max_results: int = 100,
) -> str | None:
    """Find the best matching category for a description based on existing transactions.

    Filters to top max_results recent transactions with non-null categories,
    calculates similarity, groups by category, and returns the best match
    if the weighted score exceeds 0.4.
    """
    candidates = [t for t in existing_transactions if t.get("category") is not None][
        :max_results
    ]

    if not candidates:
        return None

    # Calculate similarity for each candidate
    matches = []
    for t in candidates:
        similarity = calculate_description_similarity(description, t["description"])
        if similarity > 0.3:
            matches.append({"category": t["category"], "similarity": similarity})

    if not matches:
        return None

    # Group by category, score by (avg_similarity * 0.7 + max_similarity * 0.3)
    category_scores: dict[str, list[float]] = {}
    for m in matches:
        category_scores.setdefault(m["category"], []).append(m["similarity"])

    best_category = None
    best_score = 0.0
    for category, scores in category_scores.items():
        avg_sim = sum(scores) / len(scores)
        max_sim = max(scores)
        weighted = avg_sim * 0.7 + max_sim * 0.3
        if weighted > best_score:
            best_score = weighted
            best_category = category

    if best_score > 0.4:
        return best_category
    return None
